zfs: revamp and test compatible Kernel versions

amarshall opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

amarshall committed:
zfs: verify kernel compatibility against META

Switch to specifying the Kernel compatibility in the same form as in the
ZFS source’s `META` file. This then allows checking it against what’s in
`META` to remove the need to manually check, and so reduce the cognitive
load and change of mistakes when performing updates.

As the “max” version is indeed a compatible version, we need to
increment that for the comparison when using `versionOlder`.

One side-effect here is that it makes it not possible to override the
specified compatibility. We will “fix” this soon.

adamcstephens commented:
I’ll take a look at the few pending PRs this weekend. Sorry for the delay.

adamcstephens commented on pkgs/os-specific/linux/zfs/generic.nix:
```diff
@@ -46,7 +46,9 @@ let
-      kernelCompatible ? null,
+      kernelMinSupportedMajorMinor,
+      kernelMaxSupportedMajorMinor,
+      kernelMaxBuildableMajorMinor ? kernelMaxSupportedMajorMinor, # Only use for pre-release ZFS
```
Could we make this a boolean instead? something like `ignoreKernelCompatibility`

adamcstephens replied:
Where should we document overriding this, to allow building outside the limits?

adamcstephens commented on pkgs/os-specific/linux/zfs/generic.nix:
```diff
@@ -137,6 +155,14 @@ let
+          echo 'Checking kernelMinSupportedMajorMinor is correct...'
+          grep --quiet '^Linux-Minimum: *${lib.escapeRegex kernelMinSupportedMajorMinor}$' META
+          echo 'Checking kernelMaxSupportedMajorMinor is correct...'
+          grep --quiet '^Linux-Maximum: *${lib.escapeRegex kernelMaxSupportedMajorMinor}$' META
```
Is it fatal if this fails? If so, how does this interact with kernelMaxBuildableMajorMinor?

adamcstephens commented:
FYI, upstream is now hard failing on incompatible versions during configure. 

https://github.com/openzfs/zfs/commit/410287f7f838d967fb8546c524ca3fccad1ea9bc

amarshall committed:
zfs: add enableUnsupportedExperimentalKernel option

This will allow building ZFS against an unsupported Kernel version.
While possible, it is not supported by upstream or Nixpkgs. As such, we
don’t enable it by default, even for zfs_unstable. Enabling should be a
deliberate choice by the user to opt into potential data corruption bugs.

adamcstephens reviewed (approved):
I think we should merge this, but the conflict will need to be resolved first.

akirak commented:
@amarshall This PR has dropped support for ZFS on Linux 6.15, which has been previously working. 6.15 is officially supported on OpenZFS:

https://endoflife.date/openzfs

I think the value of `kernelMaxSupportedMajorMinor` should have been `6.16` here:

https://github.com/amarshall/nixpkgs/blob/5d2750268b5a9464354645382d5a9d63ac0ec09f/pkgs/os-specific/linux/zfs/unstable.nix#L13

adamcstephens commented:
The latest kernel is now 6.16, perhaps that's the issue?

On master:
```
𑁱 nom build -f . linuxKernel.packages.linux_6_15.zfs_2_3 --print-out-paths
Finished at 14:12:37 after 6s
/nix/store/h3wqx846byvzijr9922wq068gx48alqa-zfs-kernel-2.3.3-6.15.9

𑁱 nom build -f . linuxPackages_latest.zfs_2_3 --print-out-paths
error:
<SNIP>
       error: Package ‘zfs-kernel-2.3.3-6.16’ in /home/adam/git/nixpkgs/pkgs/os-specific/linux/zfs/generic.nix:324 is marked as broken, refusing to evaluate.
```

akirak commented:
> The latest kernel is now 6.16, perhaps that's the issue?

You are right. After updating nixpkgs, the system fails to build.

> What makes you think that? nix-instantiate --eval -A linuxKernel.packages.linux_6_15.zfs_2_3.meta.broken shows
false on current master.

Sorry about my confusion. Somehow I was unable to find out this method.
